src/meta/meta_ckpt.c

```diff
@@ -503,42 +537,224 @@ meta_blk_mods_load(WT_SESSION_IMPL *session, const char *config, WT_CKPT *ckpt,
  */
 int
 __wt_meta_ckptlist_get(
-  WT_SESSION_IMPL *session, const char *fname, bool update, WT_CKPT **ckptbasep)
+  WT_SESSION_IMPL *session, const char *fname, bool update, WT_CKPT **ckptbasep, size_t *allocated)
 {
     WT_DECL_RET;
     char *config;
 
     config = NULL;
 
     WT_ERR(__wt_metadata_search(session, fname, &config));
-    WT_ERR(__wt_meta_ckptlist_get_from_config(session, update, ckptbasep, config));
+    WT_ERR(__wt_meta_ckptlist_get_from_config(session, update, ckptbasep, allocated, config));
 
 err:
     __wt_free(session, config);
     return (ret);
 }
 
+#ifdef HAVE_DIAGNOSTIC
+/*
+ * __assert_ckpt_matches --
+ *     Assert that given two checkpoints match.
+ */
+static void
+__assert_ckpt_matches(WT_SESSION_IMPL *session, WT_CKPT *ckpt_a, WT_CKPT *ckpt_b)
+{
+    /*
+     * We are not checking checkpoint time, because there could be a minute difference depending
+     * upon when the checkpoint information was generated. This is acceptable.
+     */
+    WT_ASSERT(session,
+      (ckpt_a->name == NULL && ckpt_b->name == NULL) ||
+        (ckpt_a->name != NULL && ckpt_b->name != NULL && strcmp(ckpt_a->name, ckpt_b->name) == 0));
+    WT_ASSERT(session, ckpt_a->order == ckpt_b->order);
+    WT_ASSERT(session, ckpt_a->size == ckpt_b->size);
+    WT_ASSERT(session, ckpt_a->write_gen == ckpt_b->write_gen);
+    WT_ASSERT(session, ckpt_a->run_write_gen == ckpt_b->run_write_gen);
+    WT_ASSERT(session,
+      ckpt_a->ta.newest_start_durable_ts == ckpt_b->ta.newest_start_durable_ts &&
+        ckpt_a->ta.newest_stop_durable_ts == ckpt_b->ta.newest_stop_durable_ts &&
+        ckpt_a->ta.oldest_start_ts == ckpt_b->ta.oldest_start_ts &&
+        ckpt_a->ta.newest_txn == ckpt_b->ta.newest_txn &&
+        ckpt_a->ta.newest_stop_ts == ckpt_b->ta.newest_stop_ts &&
+        ckpt_a->ta.newest_stop_txn == ckpt_b->ta.newest_stop_txn &&
+        ckpt_a->ta.prepare == ckpt_b->ta.prepare);
+    /*
+     * The two WT_CKPT structures are created through different paths, specifically in one path the
+     * WT_CKPT.addr and WT_CKPT.raw fields are taken from a configuration file as strings including
+     * a training nul byte. Use the minimum size of the data to ignore that nul byte. Passing nul
+     * pointers to memcmp is undefined, so handle that separately.
+     */
+    WT_ASSERT(session,
+      (ckpt_a->addr.data == NULL && ckpt_b->addr.data == NULL) ||
+        (ckpt_a->addr.data != NULL && ckpt_b->addr.data != NULL &&
+          memcmp(ckpt_a->addr.data, ckpt_b->addr.data,
+            WT_MIN(ckpt_a->addr.size, ckpt_b->addr.size)) == 0));
+    WT_ASSERT(session,
+      (ckpt_a->raw.data == NULL && ckpt_b->raw.data == NULL) ||
+        (ckpt_a->raw.data != NULL && ckpt_b->raw.data != NULL &&
+          memcmp(ckpt_a->raw.data, ckpt_b->raw.data, WT_MIN(ckpt_a->raw.size, ckpt_b->raw.size)) ==
+            0));
```

Comment:
Is there a reason we don't check that .size fields match? Using `WT_MIN` here seems odd.